Bonsai.Configuration/ScriptExtensions.cs

```diff
@@ -195,6 +202,13 @@ public void AddAssemblyReferences(IEnumerable<string> assemblyReferences)
             File.WriteAllText(ProjectFileName, root.ToString(SaveOptions.DisableFormatting));
         }
 
+        public bool GetAllowUnsafeBlocks()
```

Comment:
I agree it is annoying to have to open the file multiple times just to retrieve the value of these properties, and that indeed the refactoring seems painfully non-trivial.

I guess the way I might do it is have all these methods take an `XDocument` as argument and make `LoadProjectDocument` non-static and public:

```c#
        public XDocument LoadProjectDocument()
        {
            if (!File.Exists(ProjectFileName)) return new XDocument();
            using var stream = File.OpenRead(ProjectFileName);
            using var reader = XmlReader.Create(stream, GetXmlReaderSettings());
            return XDocument.Load(reader, LoadOptions.None);
        }
```

Because the `XDocument` instance is created and loaded into memory all at once we don't have to worry about keeping track of stream dispose calls, so the call sites would just need to use `LoadProjectDocument` first and pass the document into each of the methods to get the different bits of information, e.g. for the `GetAllowUnsafeBlocks` method:

```c#
        public bool GetAllowUnsafeBlocks(XDocument document)
            => GetBoolProperty(document, AllowUnsafeBlocksElement);
```

Incidentally this also optimizes another call to `GetPackageReferences` which was already opening the file twice inside `ScriptExtensionsProvider` and which could now be refactored like so:

```c#
            var projectDocument = scriptEnvironment.LoadProjectDocument();
            using (var cacheContext = new SourceCacheContext())
            {
                var projectReferences = from id in scriptEnvironment.GetPackageReferences(projectDocument)
                                        from assemblyReference in FindAssemblyReferences(
                                            projectFramework,
                                            dependencyResource,
                                            localPackageResource,
                                            cacheContext,
                                            packageId: id)
                                        select assemblyReference;
                assemblyNames.AddRange(scriptEnvironment.GetAssemblyReferences(projectDocument));
                assemblyNames.AddRange(projectReferences);
            }
```

The call site at `DependencyInspector` would have to be refactored as well, but I guess that would be it.

Although some of these are "public" classes, the entire `Bonsai.Configuration` assembly is internalized inside the bootstrapper projects, so never visible to downstream consumers and therefore it shouldn't be an API breaking change anyway.

Let me know what you think. I didn't actually try and test this refactoring, so there might be something important I missed, for example in the assumption that creating an empty `XDocument` works for the case where the project file does not exist.

Reply:
> Although some of these are "public" classes, the entire `Bonsai.Configuration` assembly is internalized inside the bootstrapper projects, so never visible to downstream consumers and therefore it shouldn't be an API breaking change anyway.

This was an important piece of context I was definitely missing. The class being public is a lot of what made me hesitant to do anything that'd change the API or might cause subtle changes in behavior.

Originally I was also hesitating to refactor things since my gut said this all might be replaceable by the [Roslyn Workspace model](https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/work-with-workspace) when https://github.com/bonsai-rx/bonsai/issues/778 happens, but looking a bit closer I'm noticing that [Bonsai.DependencyInspector.GetPackageDependencies](https://github.com/bonsai-rx/bonsai/blob/9357b9ee76efee403a2717b225acb02307681bf2/Bonsai/DependencyInspector.cs#L118) relies on getting a list of package references and I'm not sure if that's something Roslyn can provide. (IIRC the workspace model is pretty limited as it has a very Roslyn-centric view of projects.)

Anyway, thoughts aside I think the general strategy is good for avoiding any unintended behavior changes I'll test and polish it slightly and push a commit later today.

Reply:
Make that tomorrow, https://github.com/bonsai-rx/bonsai/issues/1740 turned into a rabbit hole

Reply:
This is basically what I ended up doing except I wrapped the `XDocument` in a new type `ScriptExtensionsProjectMetadata` which takes all of the read-only project helpers.

There's definitely more elegant refactorings out there but I think it's probably better to keep things simple.